components/add.js

```diff
@@ -0,0 +1,331 @@
+   * errors that come back from the server, this list will contain the error objects
+   */
+  const [saveErrors, setSaveErrors] = useState([]);
+  const isAddable = addable(collection);
+
+  /**
+   * Interactivity properties of the underlying Ace editor. `canEdit` implies the
+   * user may modify the text field. `canSave` means that the Save button is active.
+   * errors list indicates that the JSON entered in the field is malformed. Errors
+   * is passed to the "annotations" of the underlying Ace editor which will show an
+   * indicator that there's a JSON syntax error.
+   *
+   * If there are errors the text should not be saveable. If the user has insufficient
+   * permissions both saving and editing should be disabled.
+   */
+  const [editorStatus, setEditorStatus] = useState({
+    canEdit: isAddable,
+    canSave: isAddable,
+    errors: [],
+  });
+
+  // Profile Path for the "add" action in this collection
+  const profilePath = actionProfile(collection, "add");
+
+  useEffect(() => {
+    // Grab the schema from profilePath so we can make an empty object template
+    new FetchRequest({ session }).getObject(profilePath).then((schema) => {
+      // We have the schema, so produce a dummy json from the schema
+      const biggerSchema = convertOptionalIntoRequiredSchema(schema);
+      const basic = empty(biggerSchema);
+      setText(JSON.stringify(basic, null, 4));
+    });
+  }, [profilePath, session]);
+
+  const onChange = (newValue) => {
+    setText(newValue);
+    const errors = jsonErrors(newValue);
+    const isAddable = addable(collection);
+    const status = {
+      canEdit: isAddable,
+      canSave: errors.length == 0 && isAddable,
+      errors,
+    };
+    setEditorStatus(status);
+  };
+
+  const save = () => {
+    setEditorStatus({
+      canEdit: false,
+      canSave: false,
+      errors: [],
+    });
+
+    const value = sortedJson(JSON.parse(text));
+    const collectionPath = collectionPathWithoutLimit(collection);
+    new FetchRequest({ session })
+      .postObject(collectionPath, value)
+      .then((response) => {
+        if (response.status === "success") {
+          setSaveErrors([]);
+          router.push(collectionPath);
+        } else {
+          setEditorStatus({
+            canEdit: true,
+            canSave: true,
+            errors: [],
+          });
+
+          const errors = response.errors
+            ? response.errors.map((err) => ({
+                description: err.description,
+                keys: err.name
+                  .map((val) => {
+                    return `\`${val}\``;
+                  })
+                  .join(", "),
+              }))
+            : [
+                {
+                  description:
+                    "Error saving new item, ensure the fields are filled out correctly",
+                  keys: "Generic Error",
+                },
+              ];
+          setSaveErrors(errors);
+        }
+      });
```

Comment:
We now have a `FlashMessage` component to display temporary messages on the page, like error messages. The bulleted list seems like an odd UI that should get fixed here and in IGVF-201.

<img width="904" alt="Screen Shot 2022-10-04 at 14 57 39" src="https://user-images.githubusercontent.com/1181324/193941373-e00d00b2-6e46-455a-84fa-4a2885590720.png">

Reply:
maybe we should just add this to IGVF-201?

Reply:
Sure, I think that makes sense.

Reply:
updated in IGVF-201